purchase_last_price_info/models/product.py

```diff
@@ -0,0 +1,71 @@
+
+from odoo import api, fields, models
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    last_purchase_price = fields.Float(
+        string='Last Purchase Price')
+    last_purchase_date = fields.Date(
+        string='Last Purchase Date')
+    last_supplier_id = fields.Many2one(
+        comodel_name='res.partner', string='Last Supplier')
+
+    @api.multi
+    def set_product_last_purchase(self, order_id=False):
+        """ Get last purchase price, last purchase date and last supplier """
+        PurchaseOrderLine = self.env['purchase.order.line']
+        if not self.check_access_rights('write', raise_exception=False):
+            return
+        for product in self:
+            date_order = False
+            price_unit_uom = 0.0
+            last_supplier = False
+
+            # Check if Order ID was passed, to speed up the search
+            if order_id:
+                lines = PurchaseOrderLine.search([
+                    ('order_id', '=', order_id),
+                    ('product_id', '=', product.id)], limit=1)
+            else:
+                lines = PurchaseOrderLine.search(
+                    [('product_id', '=', product.id),
+                     ('state', 'in', ['purchase', 'done'])]).sorted(
+                    key=lambda l: l.order_id.date_order, reverse=True)
+
+            if lines:
+                # Get most recent Purchase Order Line
+                last_line = lines[:1]
+
+                date_order = last_line.order_id.date_order
+                # Compute Price Unit in the Product base UoM
+                price_unit_uom = product.product_tmpl_id.uom_id.\
+                    _compute_quantity(last_line.price_unit,
+                                      last_line.product_uom)
+                last_supplier = last_line.order_id.partner_id
+
+            # Assign values to record
+            product.last_purchase_date = date_order
```

Comment:
Do you need to assign values if no lines are found

Reply:
Yes, you need this in case you cancel all Purchase Orders related to that product. In this case you need to set the values of the product to False and 0, otherwise they would stay set to the last cancelled PO values.

Reply:
Good point :+1: